//
// Aspia Project
// Copyright (C) 2020 Dmitry Chapyshev <dmitry@aspia.ru>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
//

syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package proto;

enum RouterSession
{
    ROUTER_SESSION_UNKNOWN         = 0;
    ROUTER_SESSION_ADMIN           = 1;
    ROUTER_SESSION_AUTHORIZED_PEER = 2;
    ROUTER_SESSION_ANONIMOUS_PEER  = 4;
    ROUTER_SESSION_RELAY           = 8;
}

enum ConnectionType
{
    CONNECTION_TYPE_UNKNOWN = 0;
    CONNECTION_TYPE_DIRECT  = 1;
    CONNECTION_TYPE_RELAY   = 2;
}

message EndPoint
{
    string host = 1;
    uint32 port = 2;
}

message PeerIdRequest
{
    enum Type
    {
        UNKNOWN     = 0;
        NEW_ID      = 1;
        EXISTING_ID = 2;
    }

    Type type = 1;
    bytes key = 2;
}

message PeerIdResponse
{
    fixed64 peer_id = 1;
    bytes key = 2;
}

message ConnectionRequest
{
    fixed64 peer_id = 1;
}

message ConnectionCandidate
{
    repeated EndPoint end_point = 1;
}

message ConnectionOffer
{
    ConnectionType type = 1;
    EndPoint address    = 2;
    bytes key           = 3;
}

message ConnectionResponse
{
    enum ErrorCode
    {
        UNKNOWN        = 0;
        SUCCESS        = 1;
        PEER_NOT_FOUND = 2;
        ACCESS_DENIED  = 3;
    }

    ErrorCode error_code  = 1;
    ConnectionOffer offer = 2;
}

message RouterToPeer
{
    PeerIdResponse peer_id_response        = 1;
    ConnectionRequest connection_request   = 2;
    ConnectionResponse connection_response = 3;
    ConnectionOffer connection_offer       = 4;
}

message PeerToRouter
{
    PeerIdRequest peer_id_request            = 1;
    ConnectionRequest connection_request     = 2;
    ConnectionCandidate connection_candidate = 3;
}
